process: remove redudant lazy loading from 'warnings'

RedYetiDev opened the pull request:
<!--
Before submitting a pull request, please read:

- the CONTRIBUTING guide at https://github.com/nodejs/node/blob/HEAD/CONTRIBUTING.md
- the commit message formatting guidelines at
  https://github.com/nodejs/node/blob/HEAD/doc/contributing/pull-requests.md#commit-message-guidelines

For code changes:
1. Include tests for any bug fixes or new features.
2. Update documentation if relevant.
3. Ensure that `make -j4 test` (UNIX), or `vcbuild test` (Windows) passes.

If you believe this PR should be highlighted in the Node.js CHANGELOG
please add the `notable-change` label.

Developer's Certificate of Origin 1.1

By making a contribution to this project, I certify that:

(a) The contribution was created in whole or in part by me and I
    have the right to submit it under the open source license
    indicated in the file; or

(b) The contribution is based upon previous work that, to the best
    of my knowledge, is covered under an appropriate open source
    license and I have the right under that license to submit that
    work with modifications, whether created in whole or in part
    by me, under the same open source license (unless I am
    permitted to submit under a different license), as indicated
    in the file; or

(c) The contribution was provided directly to me by some other
    person who certified (a), (b) or (c) and I have not modified
    it.

(d) I understand and agree that this project and the contribution
    are public and that a record of the contribution (including all
    personal information I submit with it, including my sign-off) is
    maintained indefinitely and may be redistributed consistent with
    this project or the open source license(s) involved.
-->

The PR removes the redundant lazy loading of several modules within the file

(Will squash when ready)

benjamingr commented:
Hey, are you sure you grok why the code is written the way it is? Why `options` is imported inline in a `lazyOption` function and why certain stuff is "lazy" loaded?

RedYetiDev commented:
> Hey, are you sure you grok why the code is written the way it is? Why `options` is imported inline in a `lazyOption` function and why certain stuff is "lazy" loaded?

The `options` loaded in the function should be the same `options` loaded at the beginning of the file. They are both references to the same object, so it (in my opinion) makes more sense to load it at the beginning. Lazy-calling the function doesn't make sense here, as nothing that would be loaded dynamically interferes with this call.

benjamingr commented:
That's likely an oversight either in the later added usage of `getOptionValue` (in onWarning), and if it's not the fix would still likely be to remove or at least rename the function

RedYetiDev commented:
> getOptionValue

Probably, I can change the PR to remove it

RedYetiDev commented:
I removed the lazy loading of `warningFile` and `fs`. The warningFile is set to undefined in a function, which is redundant, as it will always be set to the same value later on. The `fs` module doesn't need to be imported separately, so I moved it's import to the file start.

The only thing that is 'lazy' loaded is the `fd` variable.

benjamingr commented:
Do you understand _why_ these things are lazy loaded?

RedYetiDev commented:
> Do you understand _why_ these things are lazy loaded?

Unfortunately, I don't. The lazy loading seems unnecessary. BTW, It appears that the last time the function was modified was in <https://github.com/nodejs/node/commit/b55ab01b456645bcd7fce5c394f2722112080c6b>, by @refack and @BridgeAR. Let's get their opinion's on this, maybe they know why?

benjamingr commented:
@RedYetiDev or you can just ask :)

The point in lazy loading is that Node should run (ideally) as little code as possible when it boots so it can boot faster. e.g. the fs code wouldn't load in a Node program that doesn't otherwise use fs through this code path unless a warning is written to the warning file.

RedYetiDev commented:
> @RedYetiDev or you can just ask :)
> 
> The point in lazy loading is that Node should run (ideally) as little code as possible when it boots so it can boot faster. e.g. the fs code wouldn't load in a Node program that doesn't otherwise use fs through this code path unless a warning is written to the warning file.

Oh! Sorry for not asking! I can undo my last commit, which will leave the lazy loading alone, but still prevent the double-importing of the `options` module, but on the other hand, is lazy loading really needed here? Nothing else in the file is lazy-loaded

RedYetiDev commented:
I made a change to use the `getLazy` function from utils, I can undo it if you'd like, but maybe it'll work?

benjamingr reviewed (approved):
This looks fine (for warnings) lgtm. Note that in other places in the code we should consider if the cost of the extra function call is worth it (warnings + writing to a file should mean this change is fine)

RedYetiDev commented:
I'm not quite sure what went wrong in these test failures. I did some local testing, and it worked fine. Do you see any issues, @benjamingr?

benjamingr commented:
Here are some failing tests for you to look at, run these files locally:

```
such file or directory, open '/home/runner/work/node/node/test/.tmp.2469/warnings.txt'
    at ReadFileContext.<anonymous> (/home/runner/work/node/node/test/common/index.js:436:12)
    at ReadFileContext.callback (/home/runner/work/node/node/test/common/index.js:473:15)
    at FSReqCallback.readFileAfterOpen [as oncomplete] (node:fs:300:13) {
  generatedMessage: false,
  code: 'ERR_ASSERTION',
  actual: [Error: ENOENT: no such file or directory, open '/home/runner/work/node/node/test/.tmp.2469/warnings.txt'] {
    errno: -2,
    code: 'ENOENT',
    syscall: 'open',
    path: '/home/runner/work/node/node/test/.tmp.2469/warnings.txt'
  },
  expected: null,
  operator: 'ifError'
}

Node.js v22.0.0-pre
Command: out/Release/node --test-reporter=spec --test-reporter-destination=stdout --test-reporter=./tools/github_reporter/index.js --test-reporter-destination=stdout /home/runner/work/node/node/test/parallel/test-process-redirect-warnings.js
=== release test-process-redirect-warnings-env ===
Path: parallel/test-process-redirect-warnings-env
Error: --- stderr ---
(node:140104) Warning: a bad practice warning
(Use `node --trace-warnings ...` to show where the warning was created)
node:assert:1000
    throw newErr;
    ^

AssertionError [ERR_ASSERTION]: ifError got unwanted exception: ENOENT: no such file or directory, open '/home/runner/work/node/node/test/.tmp.2470/warnings.txt'
    at ReadFileContext.<anonymous> (/home/runner/work/node/node/test/common/index.js:436:12)
    at ReadFileContext.callback (/home/runner/work/node/node/test/common/index.js:473:15)
    at FSReqCallback.readFileAfterOpen [as oncomplete] (node:fs:300:13) {
  generatedMessage: false,
  code: 'ERR_ASSERTION',
  actual: [Error: ENOENT: no such file or directory, open '/home/runner/work/node/node/test/.tmp.2470/warnings.txt'] {
    errno: -2,
    code: 'ENOENT',
    syscall: 'open',
    path: '/home/runner/work/node/node/test/.tmp.2470/warnings.txt'
  },
  expected: null,
  operator: 'ifError'
}

```

RedYetiDev commented:
Found the issue

MoLow commented:
@RedYetiDev test failures still seem related. please run tests locally as [recommended in the contribution guide](https://github.com/nodejs/node/blob/main/doc/contributing/pull-requests.md#step-6-test) prior to opening a PR

RedYetiDev commented:
> @RedYetiDev test failures still seem related. please run tests locally as [recommended in the contribution guide](https://github.com/nodejs/node/blob/main/doc/contributing/pull-requests.md?rgh-link-date=2024-04-13T21%3A15%3A08Z#step-6-test) prior to opening a PR

Sorry! I'll take a look over the coming days.

RedYetiDev commented:
If it aint broke don't fix it :-)
